modules/ipc/include/hephaestus/ipc/zenoh/action_server.h

```diff
@@ -0,0 +1,254 @@
+};
+
+template <typename StatusT>
+using StatusUpdateCallback = std::function<void(const StatusT&)>;
+
+/// Call the action server with the given request.
+/// - The status_update_cb will be called with the status updates from the server.
+///   - The updates are decided by the user implementation of the action server.
+/// - Returns a future which will eventually contain the response from the server.
+template <typename RequestT, typename StatusT, typename ReplyT>
+[[nodiscard]] auto callActionServer(SessionPtr session, const TopicConfig& topic_config,
+                                    const RequestT& request, StatusUpdateCallback<StatusT>&& status_update_cb)
+    -> std::future<ActionServerResponse<ReplyT>>;
+
+/// Request the action server to stop.
+[[nodiscard]] auto requestActionServerToStopExecution(Session& session,
+                                                      const TopicConfig& topic_config) -> bool;
+
+// TODO: add a function to get notified when the server is idle again.
+
+// ----------------------------------------------------------------------------------------------------------
+// --------- ActionServer Implementation --------------------------------------------------------------------
+// ----------------------------------------------------------------------------------------------------------
+
+template <typename RequestT, typename StatusT, typename ReplyT>
+ActionServer<RequestT, StatusT, ReplyT>::ActionServer(SessionPtr session, TopicConfig topic_config,
+                                                      ActionTriggerCallback&& action_trigger_cb,
+                                                      ExecuteCallback&& execute_cb)
+  : session_(std::move((session)))
+  , topic_config_(std::move(topic_config))
+  , action_trigger_cb_(std::move(action_trigger_cb))
+  , execute_cb_(std::move(execute_cb))
+  , request_service_(std::make_unique<Service<RequestT, ActionServerRequestResponse>>(
+        session_, topic_config_, [this](const RequestT& request) { return onRequest(request); }))
+  , request_consumer_([this](const RequestT& request) { return execute(request); }, std::nullopt) {
+  request_consumer_.start();
+}
+
+template <typename RequestT, typename StatusT, typename ReplyT>
+ActionServer<RequestT, StatusT, ReplyT>::~ActionServer() {
+  request_consumer_.stop();
+}
+
+template <typename RequestT, typename StatusT, typename ReplyT>
+auto ActionServer<RequestT, StatusT, ReplyT>::onRequest(const RequestT& request)
+    -> ActionServerRequestResponse {
+  if (is_running_) {
+    LOG(ERROR) << fmt::format("ActionServer (topic: {}): server is already serving one request.",
+                              topic_config_.name);
+
+    return { .status = ActionServerRequestStatus::REJECTED_ALREADY_RUNNING };
+  }
+
+  try {
+    const auto response = action_trigger_cb_(request);
+    if (response != ActionServerTriggerStatus::SUCCESSFUL) {
+      return { .status = ActionServerRequestStatus::REJECTED_USER };
```

Comment:
What is the difference between REJECTED_USER and INVALID?

Reply:
For example if you call a service with wrong data (of different type) it would be rejected before it reaches the user